### PR TITLE
FEATURE: Display PMs in assigned activity and differentiate them.

### PR DESCRIPTION
**Pull request by romanrizzi**
Screen Shot 2019-04-18 at 15 28 00

### CODE REVIEW

File: assets/javascripts/discourse-assign/controllers/user-activity-assigned.js.es6 [hunk 1]
```diff
@@ -21,10 +23,25 @@ export default Ember.Controller.extend({
...
               type: "PUT",
               data: { user_id: user.get("id") }
-            }).then(() => this.send("unassignedAll"));
+            }).then(() => this.send("changeAssigned"));
           }
         }
       );
+    },
+    unassign(topic) {
+      ajax("/assign/unassign", {
```
Comment (SamSaffron) [inline assets/javascripts/discourse-assign/controllers/user-activity-assigned.js.es6]: hmmm isn't this ajax call elsewhere, should we not be importing a model of sorts and just calling it?

File: assets/javascripts/discourse/templates/components/assigned-topic-list-item.hbs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+{{#if bulkSelectEnabled}}
+  <td class="bulk-select">
```
Comment (SamSaffron) [inline assets/javascripts/discourse/templates/components/assigned-topic-list-item.hbs]: do we need bulk select here?

File: assets/javascripts/discourse/components/assigned-topic-list-item.js.es6 [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  },
+
+  @computed("topic.archetype")
+  isPrivateMessage(archetype) {
+    return archetype === "private_message";
+  },
+
+  actions: {
+    toggleBookmark() {
+      this.get("topic").toggleBookmark();
```
Comment (SamSaffron) [inline assets/javascripts/discourse/components/assigned-topic-list-item.js.es6]: do we need this in the list?

File: assets/javascripts/discourse/components/assigned-topic-list-item.js.es6 [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  }
+}
+
+export function navigateToTopic(topic, href) {
+  this.appEvents.trigger("header:update-topic", topic);
+  DiscourseURL.routeTo(href || topic.get("url"));
+  return false;
+}
+
+export default Ember.Component.extend({
```
Comment (SamSaffron) [inline assets/javascripts/discourse/components/assigned-topic-list-item.js.es6]: Can we not just inherit off core here and reuse the majority of the implementation?

**Comment by SamSaffron** [comment]
My overriding concern here is that we are cutting and pasting a lot of code from core, I wonder if there are better ways of reusing bits, @eviltrout any ideas here?

**Comment by eviltrout** [comment]
Yes when @romanrizzi mentioned this on our call I thought it was just the template being copied over, but here I see there's a lot of Javascript code too.

Could we not use the component that renders the topic list and use a plugin outlet for the assigned specific controls? Or at least extend it like sam mentions?

**Comment by romanrizzi** [comment]
I came up with a different approach here:

https://github.com/discourse/discourse/pull/7433

That PR aims to decouple the default behavior of a `topic-list-item` from the component itself, making it easy to re-use it from another component defining a non-raw version of that component.

This way we can rewrite the `assign-topic-list-item` component as: 

```javascript
Ember.Component.extend(ListItemDefaults)
```

**Comment by SamSaffron** [comment]
@eviltrout when you get back can you take care of merging this + the core PR ?

**Comment by tgxworld** [comment]
Does this replace the filter I added for messages?

Screenshot from 2019-04-29 16-02-50

**Comment by romanrizzi** [comment]
@tgxworld The messages section uses a different route and will remain unchanged.

If we merge the code PR, we need to be aware that it won't be available until `beta10`. What's the best strategy to avoid breaking instances running `beta9` and `discourse-assign`?

**Comment by eviltrout** [comment]
@romanrizzi I've approved the core PR so I presume this now needs to be updated?

**Comment by romanrizzi** [comment]
@eviltrout Done!

**Comment by romanrizzi** [comment]
I was thinking about this yesterday.

I don't know if this is possible in our codebase but I think we can check if the Mixin is present using the `require` primitive and then decide what we're gonna do.

We can ship a minimal `ListItemDefault` object and fallback to it if necessary (and raise a warning).

 Versioning this plugin would be awesome to avoid shipping breaking changes because we depend on core's code.

**Comment by eviltrout** [comment]
@romanrizzi I do believe in the past we looked at pinning plugins to branches. I would look into that and see if we are still doing it (for example stable / beta.) If not the require solution might work. You could also consider comitting an empty version of the mixin to the stable/beta branches so the import does not fail, but check if the object is empty.

Feel free to discuss on dev / meta if you can find a topic about plugin/version maintenance.

**Comment by SamSaffron** [comment]
Thanks Roman, trying this out!

**Pull request merged**